### DISCUSSION TITLE
How to add attribute to tag in navit.xml?

### DISCUSSION THREAD

**Discussion by adevoss** [navit-gps/navit#1258] [Q&A] [answered]
For a new PR I want to add an attribute in the `speech` tag in `navit.xml`.
How is this done?
I added the attribute in `attr_def.h` and put an extra property in `struct speech_priv`.
Same as the attribute `sample_dir`.

This does not work. What am I missing?

**Comment by mvglasow** [comment]
If I remember correctly... when `navit.xml` is read, attributes found in there are passed to the attribute setter functions for the appropriate module. Make sure the attribute setter function for `speech` knows what to do with this attribute.

First, search `speech.c` for its `struct object_func` and examine its `set_attr` and `add_attr` members. Some modules use a generic function while others define their own – off the top of my head I’m not sure which is the case for `speech`.

Then, make sure this function does whatever you want it to when it is called with your attribute. If you are storing an attribute in a `struct` member, the module in question will need its own method (which can still fall back to the generic one for any attribute it doesn’t handle).

Next, you might also need to ensure `get_attr` correctly returns your new attribute, and iterators retrieved with `iter_new` will include it.

You can look at some of the other core modules to see how this is done there. (The `navit` element in `navit.xml` and each of its children correspond to a core module, each with a corresponding `.c` source file.)

For further reading and reference (archived from the old wiki):
https://web.archive.org/web/20210227183021/https://wiki.navit-project.org/index.php/Core_Module – explains the anatomy of a core module
https://web.archive.org/web/20210227185807/https://wiki.navit-project.org/index.php/Attributes – about introducing attributes

(These articles should really be preserved on Read the Docs – feel free to copy the information over and enrich it with anything you feel is missing.)

**Reply by adevoss** [reply]
The new attribute can be read in the code but now I can't enter a city name when I want set a destination.
That's why I thought I did something wrong earlier and started this discussion.

The value of the new attribute is now hardcoded in #1259.
The attribute value is a TTS command.

**Reply by adevoss** [reply]
Instead of hardcoded now read from txt file.

**Comment by adevoss** [comment, marked as answer]
Solved by @OLFDB. Thanks. See #1260
